File: Enzo.Music/Canzone/Strofa.cs
using System.Numerics;

namespace Enzo.Music;

public class Strofa : ICloneable, IEquatable<Strofa>, IEqualityOperators<Strofa, Strofa, bool>
{
    public string? Nome;
    public List<Parte?> Parti = new();

    protected void CopyFrom(Strofa strofa)
    {
        Nome = strofa.Nome;
        foreach (Parte? p in strofa.Parti)
        {
            Parti.Add((Parte?)p?.Clone());
        }
    }

    public virtual object Clone()
    {
        Strofa n = new();
        n.CopyFrom(this);
        return n;
    }

    public bool Equals(Strofa? other)
    {
        if (other is null) return false;
        bool uguali = string.Equals(Nome, other.Nome) && Parti.Count == other.Parti.Count;
        if (uguali)
        {
            uguali = !Parti.Where((p, i) => p is not null && !p.Equals(other.Parti[i])).Any();
        }
        return uguali;
    }
    public override bool Equals(object? obj)
    {
        if (obj is null) return false;
        if (obj is not Strofa) return false;

        return Equals((Strofa)obj);
    }
    public static new bool Equals(object? obj1, object? obj2)
    {
        if (!(obj1 is not null && obj2 is not null)) return false;
        if (obj1 is not Strofa) return false;
        return ((Strofa)obj1).Equals(obj2);
    }

    public override int GetHashCode()
    {
        throw new NotImplementedException();
    }

    public static bool operator ==(Strofa? left, Strofa? right)
    {
        return Strofa.Equals(left, right);
    }

    public static bool operator !=(Strofa? left, Strofa? right)
    {
        return !Strofa.Equals(left, right);
    }
}

File: KantaroBLL/Canzone/Strofa.vb
Imports Enzo.Music

Public Class Strofa
    Implements ICloneable

    Public Property Nome As String
    Public Property Parti As New Generic.List(Of Parte)

    Public Overridable Function Clone() As Object Implements ICloneable.Clone
        Dim n As New Strofa
        n.CopyFrom(Me)
        Return n
    End Function

    Protected Sub CopyFrom(strofa As Strofa)
        Nome = strofa.Nome
        For Each p In strofa.Parti
            If p IsNot Nothing Then
                Parti.Add(p.Clone)
            Else
                Parti.Add(Nothing)
            End If
        Next
    End Sub

    Public Overrides Function Equals(other As Object) As Boolean
        If other Is Nothing Then Return False
        If Not TypeOf other Is Strofa Then Return False
        Dim b = DirectCast(other, Strofa)
        Dim uguali = String.Equals(Me.Nome, b.Nome) AndAlso
                     Me.Parti.Count = b.Parti.Count
        If uguali Then
            For i = 0 To Parti.Count - 1
                uguali = If(Parti(i)?.Equals(b.Parti(i)), False)
                If Not uguali Then Exit For
            Next
        End If
        Return uguali
    End Function

    Public Overloads Shared Function Equals(obj1 As Object, obj2 As Object) As Boolean
        If obj1 Is Nothing AndAlso obj2 Is Nothing Then Return True
        If Not TypeOf obj1 Is Strofa Then Return False
        Return DirectCast(obj1, Strofa).Equals(obj2)
    End Function

    Public Shared Operator =(a As Strofa, b As Strofa) As Boolean
        Return Strofa.Equals(a, b)
    End Operator

    Public Shared Operator <>(a As Strofa, b As Strofa) As Boolean
        Return Not Strofa.Equals(a, b)
    End Operator
End Class

Public Class StrofaRipetuta
    Inherits Strofa

    ''' <summary>
    ''' Nome della strofa da ripetere
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Property Riferimento As String

    Public Overrides Function Clone() As Object
        Dim c As New StrofaRipetuta
        c.CopyFrom(Me)
        c.Riferimento = Me.Riferimento
        Return c
    End Function
End Class
